python-checks/src/main/java/org/sonar/python/checks/ConfusingTypeCheckingCheck.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import javax.annotation.Nullable;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.types.InferredType;
+
+@Rule(key = "S5864")
+public class ConfusingTypeCheckingCheck extends PythonSubscriptionCheck {
+  @Override
+  public void initialize(Context context) {
+    new NonCallableCalledCheck().initialize(context);
+  }
+
+  private static class NonCallableCalledCheck extends NonCallableCalled {
+
+    @Override
+    public boolean isNonCallableType(InferredType type) {
+      return type.canHaveMember("__call__") && !type.declareMember("__call__");
```

Comment:
I'm a bit confused by `type.canHaveMember("__call__")`. Is this a way of asserting that we're dealing with a `DeclaredType`? It feels a bit indirect and implementation dependent to me.

That said I'm not sure I have a suggestion for a more direct check (checking directly that `type` is an instance of `DeclaredType` should do the trick but it might be heavy-handed...but then again this rule is tailored for `DeclaredType` only so it might make sense)

Reply:
Yes, the idea is to avoid raising twice the same issue (one for the bug rule and one for code smell).

I think we cannot check only for `type` being instance of `DeclaredType`. It could also be a `UnionType` containing `DeclaredType` and we still want to raise an issue.
I agree it's a bit indirect; if you agree I'll add a comment explaining this